File: sol/solution/0001-0100/0039. Combination Sum/Solution.ts
function combinationSum(candidates: number[], target: number): number[][] {
	const result: number[][] = [];
	function backtrack(start: number, current: number[], remaining: number): void {
		if (remaining === 0) {
			result.push([...current]);
			return;
		}
		for (let i = start; i < candidates.length; i++) {
			if (candidates[i] <= remaining) {
				current.push(candidates[i]);
				backtrack(i, current, remaining - candidates[i]);
				current.pop();
			}
		}
	}
	backtrack(0, [], target);
	return result;
}

File: sol/solution/0001-0100/0039. Combination Sum/Solution.rs
impl Solution {
    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut candidates = candidates;
        candidates.sort_unstable();
        let mut result = Vec::new();
        let mut current = Vec::new();
        fn backtrack(
            candidates: &Vec<i32>,
            start: usize,
            remaining: i32,
            current: &mut Vec<i32>,
            result: &mut Vec<Vec<i32>>,
        ) {
            if remaining == 0 {
                result.push(current.clone());
                return;
            }
            for i in start..candidates.len() {
                let num = candidates[i];
                if num > remaining {
                    break;
                }
                current.push(num);
                backtrack(candidates, i, remaining - num, current, result);
                current.pop();
            }
        }
        backtrack(&candidates, 0, target, &mut current, &mut result);
        result
    }
}